File: Algorithms/CPP/56 Merge Intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), [](vector<int>& x, vector<int>& y){
            return x[0] < y[0];
        });

        vector<vector<int>> result;
        for (auto i: intervals){
            if (result.empty()){
                result.push_back(i);
            }
            else{
                auto v = result.back();
                if (i[0] <= v[1]){
                    // merge
                    vector<int> merge_v = {v[0], max(v[1], i[1])};
                    result.pop_back();
                    result.push_back(merge_v);
                }
                else{
                    result.push_back(i);
                }
            }
        }
        return result;
    }
};

File: Algorithms/Python/56 Merge Intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda x: x[0])
        result = []
        for i in intervals:
            if len(result) == 0:
                result.append(i)
            else:
                if i[0] <= result[-1][1]:
                    # can merge
                    result[-1] = [result[-1][0], max(result[-1][1], i[1])]
                else:
                    result.append(i)
        return result
